### PR TITLE
update ActionBase._low_level_execute_command to honor executable

### LINKED ISSUE #68054: raw no longer honoring executable (2.9 vs 2.7)
##### SUMMARY

In Ansible 2.7, I was able to specify an alternative executable at the task level when running the raw module.  This behaviour has changed in 2.9; the task no longer honors executable.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

raw

##### ANSIBLE VERSION

```paste below
ansible 2.9.4
  config file = /home/rowagn/data-platform/ansible_bootstrap/ansible.cfg
  configured module search path = [u'/home/rowagn/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /sso/sfw/virtualenv/ansible29/lib/python2.7/site-packages/ansible-2.9.4-py2.7.egg/ansible
  executable location = /sso/sfw/virtualenv/ansible29/bin/ansible
  python version = 2.7.17 (default, Feb  6 2020, 10:54:17) [GCC 5.4.0 20160609]

```

##### CONFIGURATION

```paste below
ALLOW_WORLD_READABLE_TMPFILES(/home/rowagn/data-platform/ansible_bootstrap/ansible.cfg) = True
ANSIBLE_PIPELINING(/home/rowagn/data-platform/ansible_bootstrap/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/home/rowagn/data-platform/ansible_bootstrap/ansible.cfg) = -C -o ControlMaster=no -o StrictHostKey
DEFAULT_EXECUTABLE(/home/rowagn/data-platform/ansible_bootstrap/ansible.cfg) = /bin/sh
DEFAULT_HOST_LIST(/home/rowagn/data-platform/ansible_bootstrap/ansible.cfg) = [u'/home/rowagn/data-platform/ansible_
RETRY_FILES_ENABLED(/home/rowagn/data-platform/ansible_bootstrap/ansible.cfg) = False

```

##### OS / ENVIRONMENT

I'll elaborate below, but our users are not allowed to execute /bin/sh (or other shells) as sudo.

##### STEPS TO REPRODUCE

create a user (i.e., testuser) on the target machine with the following restrictions in /etc/sudoers:
```
Cmnd_Alias      SHELLS = /bin/sh
testuser  ALL=(ALL)  ALL, !SHELLS
```
create /etc/ansible-wrapper on the target machine with 755 permissions containing:
```
#!/bin/bash
[[ -z $1 ]] && exit 0
/bin/bash -c "$@"
```
create ansible.cfg containing:
```
executable=/bin/sh
```
Then run the playbook in 2.7 and 2.9 environments.
```
ansible-playbook a.yml -K -k -utestuser
```

```yaml
$ cat a.yml 
---
- hosts: all
  gather_facts: no
  tasks:
  - name: whoami
    raw: echo "test" > /tmp/test 
    args:
      executable: "/etc/ansible-wrapper"
    become: yes
    become_method: sudo
```

##### EXPECTED RESULTS

Expected (this is what happens on ansible 2.7):
```
$ ansible-playbook a.yml -K -k -utestuser -vvv
ansible-playbook 2.7.0
  config file = /home/rowagn/data-platform/ansible_bootstrap/ansible.cfg
  configured module search path = [u'/home/rowagn/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /sso/sfw/python2/lib/python2.7/site-packages/ansible-2.7.0-py2.7.egg/ansible
  executable location = /sso/sfw/python2/bin/ansible-playbook
  python version = 2.7.12 (default, Apr 25 2017, 09:26:15) [GCC 5.4.0 20160609]
Using /home/rowagn/data-platform/ansible_bootstrap/ansible.cfg as config file
SSH password: 
SUDO password[defaults to SSH password]: 
Parsed /home/rowagn/data-platform/ansible_bootstrap/hosts inventory source with ini plugin

PLAYBOOK: a.yml ******************************************************************************************************
1 plays in a.yml

PLAY [all] ***********************************************************************************************************
META: ran handlers

TASK [whoami] ********************************************************************************************************
task path: /home/rowagn/data-platform/ansible_bootstrap/a.yml:5
<rowagn-tower-test01.vsp.sas.com> ESTABLISH SSH CONNECTION FOR USER: testuser
<rowagn-tower-test01.vsp.sas.com> SSH: EXEC sshpass -d9 ssh -C -o ControlMaster=no -o StrictHostKeyChecking=no -o User=testuser -o ConnectTimeout=10 -tt rowagn-tower-test01.vsp.sas.com '/etc/ansible-wrapper -c '"'"'sudo -H -S  -p "[sudo via ansible, key=ftbmhhvbpkovfeymzyhwqpmtjrwljdkp] password: " -u root /etc/ansible-wrapper -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-ftbmhhvbpkovfeymzyhwqpmtjrwljdkp; echo "test" > /tmp/test'"'"'"'"'"'"'"'"''"'"''
Escalation succeeded
<rowagn-tower-test01.vsp.sas.com> (0, '\r\n', 'Connection to rowagn-tower-test01.vsp.sas.com closed.\r\n')
changed: [rowagn-tower-test01.vsp.sas.com] => {
    "changed": true, 
    "rc": 0, 
    "stderr": "Connection to rowagn-tower-test01.vsp.sas.com closed.\r\n", 
    "stderr_lines": [
        "Connection to rowagn-tower-test01.vsp.sas.com closed."
    ], 
    "stdout": "\r\n", 
    "stdout_lines": [
        ""
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ***********************************************************************************************************
rowagn-tower-test01.vsp.sas.com : ok=1    changed=1    unreachable=0    failed=0   

```
If I tail /var/log/secure on the target, I see the following is executed:
```
Mar  5 19:38:06 rowagn-tower-test01 sudo: testuser : TTY=pts/1 ; PWD=/home/testuser ; USER=root ; COMMAND=/etc/ansible-wrapper -c echo BECOME-SUCCESS-blsgdxudsexamxfpsscdlgkglhynitnw; echo "test" > /tmp/test
```

##### ACTUAL RESULTS

Actual (this is what happens on Ansible 2.9):

```
$ ansible-playbook a.yml -K -k -utestuser -vvv
ansible-playbook 2.9.4
  config file = /home/rowagn/data-platform/ansible_bootstrap/ansible.cfg
  configured module search path = [u'/home/rowagn/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /sso/sfw/virtualenv/ansible29/lib/python2.7/site-packages/ansible-2.9.4-py2.7.egg/ansible
  executable location = /sso/sfw/virtualenv/ansible29/bin/ansible-playbook
  python version = 2.7.17 (default, Feb  6 2020, 10:54:17) [GCC 5.4.0 20160609]
Using /home/rowagn/data-platform/ansible_bootstrap/ansible.cfg as config file
SSH password: 
BECOME password[defaults to SSH password]: 
host_list declined parsing /home/rowagn/data-platform/ansible_bootstrap/hosts as it did not pass its verify_file() method
script declined parsing /home/rowagn/data-platform/ansible_bootstrap/hosts as it did not pass its verify_file() method
auto declined parsing /home/rowagn/data-platform/ansible_bootstrap/hosts as it did not pass its verify_file() method
Parsed /home/rowagn/data-platform/ansible_bootstrap/hosts inventory source with ini plugin

PLAYBOOK: a.yml ****************************************************************************************************
1 plays in a.yml

PLAY [all] *********************************************************************************************************
META: ran handlers

TASK [whoami] ******************************************************************************************************
task path: /home/rowagn/data-platform/ansible_bootstrap/a.yml:5
<rowagn-tower-test01.vsp.sas.com> ESTABLISH SSH CONNECTION FOR USER: testuser
<rowagn-tower-test01.vsp.sas.com> SSH: EXEC sshpass -d8 ssh -C -o ControlMaster=no -o StrictHostKeyChecking=no -o 'User="testuser"' -o ConnectTimeout=10 -tt rowagn-tower-test01.vsp.sas.com '/etc/ansible-wrapper -c '"'"'sudo -H -S  -p "[sudo via ansible, key=iayqkqennnjnowmcjazgysiwuozpbzmy] password:" -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-iayqkqennnjnowmcjazgysiwuozpbzmy ; echo "test" > /tmp/test'"'"'"'"'"'"'"'"''"'"''
Escalation succeeded
<rowagn-tower-test01.vsp.sas.com> (1, '\r\n', 'Connection to rowagn-tower-test01.vsp.sas.com closed.\r\n')
<rowagn-tower-test01.vsp.sas.com> Failed to connect to the host via ssh: Connection to rowagn-tower-test01.vsp.sas.com closed.
fatal: [rowagn-tower-test01.vsp.sas.com]: FAILED! => {
    "changed": true, 
    "msg": "non-zero return code", 
    "rc": 1, 
    "stderr": "Connection to rowagn-tower-test01.vsp.sas.com closed.\r\n", 
    "stderr_lines": [
        "Connection to rowagn-tower-test01.vsp.sas.com closed."
    ], 
    "stdout": "\r\n", 
    "stdout_lines": [
        ""
    ]
}

PLAY RECAP *********************************************************************************************************
rowagn-tower-test01.vsp.sas.com : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```
If I tail /var/log/secure on the target, I see the following is executed:
```
Mar  5 19:38:32 rowagn-tower-test01 sudo: testuser : command not allowed ; TTY=pts/1 ; PWD=/home/testuser ; USER=root ; COMMAND=/bin/sh -c echo BECOME-SUCCESS-gurlxipwqzptgdiqwmrmaevflvziiiff ; echo "test" > /tmp/test
```

So in the 2.7 case, /etc/ansible-wrapper was used as the executable, as specified in the playbook, but in the 2.9 case, /bin/sh (from ansible.cfg) is used.

Here's why this is a problem for us.  Our IT crew doesn't allow us to execute shells (/bin/sh, /bin/bash, etc) with sudo.  As a workaround, we create the above wrapper script on all our target hosts and use executable=/etc/ansible-wrapper in our ansible.cfg files.  This all works fine.  However, we have a playbook that we use to bootstrap our target hosts.  The bootstrap playbook uses executable=/bin/sh (in ansible.cfg).  It copies /etc/ansible-wrapper out to the target in the user's home directory (which uses /bin/sh, since no sudo is needed).  Then, it uses the raw module to move ansible-wrapper to /etc, but since this requires sudo, it specifies /home/user/ansible-wrapper as the executable for that task (similarly to the above playbook).  Without being able to specify executable, this playbook breaks in Ansible 2.9.

### PR DESCRIPTION
**Pull request by rwagnergit**
##### SUMMARY

Fixes #68054

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ActionBase

##### ADDITIONAL INFORMATION

```
Before:

(venv) rowagn@localhost:~/testcase$ ansible-playbook a.yml -i ./hosts -K -k -utestuser
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or
trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
SSH password: 
BECOME password[defaults to SSH password]: 

PLAY [all] ******************************************************************************************************************************************

TASK [whoami] ***************************************************************************************************************************************
fatal: [rowagn-tower-test01.vsp.sas.com]: FAILED! => {"changed": true, "msg": "non-zero return code", "rc": 1, "stderr": "Shared connection to rowagn-tower-test01.vsp.sas.com closed.\r\n", "stderr_lines": ["Shared connection to rowagn-tower-test01.vsp.sas.com closed."], "stdout": "\r\n", "stdout_lines": [""]}

PLAY RECAP ******************************************************************************************************************************************
rowagn-tower-test01.vsp.sas.com : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   



After:

(venv) rowagn@localhost:~/testcase$ ansible-playbook a.yml -i ./hosts -K -k -utestuser
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or
trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
SSH password: 
BECOME password[defaults to SSH password]: 

PLAY [all] ******************************************************************************************************************************************

TASK [whoami] ***************************************************************************************************************************************
changed: [rowagn-tower-test01.vsp.sas.com]

PLAY RECAP ******************************************************************************************************************************************
rowagn-tower-test01.vsp.sas.com : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

### CODE REVIEW

**Comment by gundalow** [comment]
Thanks for this fix.
Could you please add test to ensure this doesn't regress in the future?

**Comment by rwagnergit** [comment]
@gundalow - Yes.  I may need to reach out to you for some guidance re: shippable.

**Comment by rwagnergit** [comment]
@gundalow - The following testcase is valid when I run it from the ansible-playbook command line (i.e., it fails unless the code change in this PR is incorporated):
```
- hosts: all
  gather_facts: no
  tasks:
  - name: use sh
    raw: echo $0
    args:
      executable: "/bin/sh"
    become: true
    become_method: sudo
    register: sh_output
  - name: assert sh
    assert:
      that: "(sh_output.stdout | trim) == '/bin/sh'"
  - name: use bash
    raw: echo $0
    args:
      executable: "/bin/bash"
    become: true
    become_method: sudo
    register: bash_output
  - name: assert bash
    assert:
      that: "(bash_output.stdout | trim) == '/bin/bash'"
```

However, if I add the following to test/integration/targets/raw/tasks/main.yml, it doesn't fail when I run 
```
ansible-test integration --docker ubuntu1804 -v raw
```

Any idea why the integration test isn't behaving like ansible-playbook?

**Comment by sivel** [comment]
I believe this is due to a testing implementation detail.

These tests are run in a container with the local connection, and the tests are already running as `root`.  By default `become` will be ignored when the `remote_user` and `become_user` are the same.

The structure of the raw tests would have to be rewritten to make use of a `runme.sh` that calls a playbook, that allows you to set `ANSIBLE_BECOME_ALLOW_SAME_USER=1`.

Look at the `runme.sh` in `remote_tmp` for an example, which would just have to be copied and modified to use `ANSIBLE_BECOME_ALLOW_SAME_USER`.  The `tasks/main.yml` would need moved to `playbook.yml` and turned into a playbook, that roughly follows that in `remote_tmp`.

**Comment by rwagnergit** [comment]
@sivel - that's brilliant (though I have no clue what ANSIBLE_BECOME_ALLOW_SAME_USER is).  Added a new test at test/integration/targets/raw_executable/ that does mirror what I see with my standalone ansible-playbook test.

**Comment by rwagnergit** [comment]
```
ansible-test integration --docker $E -v raw_executable
```
succeeds for all E in: { centos6, centos7, centos8, fedora30, fedora31, opensuse15py2, opensuse15, ubuntu1604, ubuntu1804 }

**Comment by rwagnergit** [comment]
Looks like the FreeBSD test failed because bash is not installed:
```
fatal: [testhost]: FAILED! => {
    "changed": true, 
    "msg": "non-zero return code", 
    "rc": 127, 
    "stderr": "/bin/sh: /bin/bash: not found\n", 
    "stderr_lines": [
        "/bin/sh: /bin/bash: not found"
    ], 
    "stdout": "", 
    "stdout_lines": []
}
```

And AIX failed because sudo doesn't exist:
```
fatal: [testhost]: FAILED! => {
    "changed": true, 
    "msg": "non-zero return code", 
    "rc": 127, 
    "stderr": "/bin/sh: sudo:  not found\n", 
    "stderr_lines": [
        "/bin/sh: sudo:  not found"
    ], 
    "stdout": "", 
    "stdout_lines": []
}
```

Not sure how to work around this.  Any suggestions?

**Comment by rwagnergit** [comment]
@sivel @gundalow I'm not familiar with the "core review" label.  What are the next steps to getting this merged?

**Comment by sivel** [comment]
@rwagnergit This PR will be triaged soon.  At some point in the future it will be fully reviewed.  Please be patient as we are in the process of a large overhaul in this repository as we prepare for the Ansible 2.10 release.

**Comment by rwagnergit** [comment]
@sivel Thanks.  Once it's merged, I'll work on backporting to 2.9.

**Comment by rwagnergit** [comment]
@sivel - I just received a notice from Red Hat that Ansible 2.7 is EOL June 18.  To that end, we really need to upgrade, but we can't until this PR is merged.  Any idea when this will move forward?  Thanks

**Comment by sivel** [comment]
This PR will need to be reviewed by someone other than myself, since I wrote the code you implemented.

I'll request a few reviews from others.

File: lib/ansible/plugins/action/__init__.py [hunk 1]
```diff
@@ -1045,6 +1045,10 @@ def _low_level_execute_command(self, cmd, sudoable=True, in_data=None, executabl
             display.debug("_low_level_execute_command(): changing cwd to %s for this command" % chdir)
             cmd = self._connection._shell.append_command('cd %s' % chdir, cmd)
 
+        # https://github.com/ansible/ansible/issues/68054
+        if executable:
```
Comment (mkrizek) [inline lib/ansible/plugins/action/__init__.py:1049]: https://github.com/ansible/ansible/blob/eaf648e60093505d47b2918e58453fe9181ea517/lib/ansible/plugins/action/__init__.py#L1063-L1070
The above code makes me wonder if this should too only be done `if self._connection.allow_executable`?

File: test/integration/targets/raw/tasks/main.yml [hunk 2]
```diff
@@ -81,3 +81,27 @@
           - 'raw_result2.stdout_lines is defined'
           - 'raw_result2.rc == 0'
           - 'raw_result2.stdout_lines == ["foobar"]'
+# the following five tests added to test https://github.com/ansible/ansible/pull/68315
```
Comment (mkrizek) [inline test/integration/targets/raw/tasks/main.yml:84]: These are just two tests using five tasks :)

File: lib/ansible/plugins/action/__init__.py (hunk 1 above)
Comment (mkrizek) [reply lib/ansible/plugins/action/__init__.py:1049]: After looking further through the code, the same seems to be already handled in `_build_success_command()` by `noexe` arg.

**Review by mkrizek — APPROVED** [review]

**Comment by rwagnergit** [comment]
Thanks @mkrizek for reviewing.

**Comment by rwagnergit** [comment]
@bcoca - would you mind kindly reviewing this?  I'm under the gun to get this merged and backported to 2.9 before 2.7 goes EOL.  Thanks much!

File: lib/ansible/plugins/action/__init__.py [hunk 3]
```diff
@@ -1045,6 +1045,10 @@ def _low_level_execute_command(self, cmd, sudoable=True, in_data=None, executabl
...
+            self._connection._shell.executable = executable
```
Comment (bcoca) [inline lib/ansible/plugins/action/__init__.py:1050]: never assign directly, use set_option(), in this case `direct={'executable'=executable} `

right now this is 'safe' due to using multi procs, we should add note about not being thread safe

**Comment by bcoca** [comment]
fix is ok for now, but this should really not be an issue with become wrapping, I need to think about a better solution.

**Review by bcoca — APPROVED** [review]

**Comment by rwagnergit** [comment]
Thanks @bcoca !

**Comment by rwagnergit** [comment]
@sivel Hey Matt - what's left?

**Comment by sivel** [comment]
This has been merged for the 2.10 release. Please feel free to create a backport for stable-2.9

**Pull request merged**